python/ray/dashboard/modules/virtual_cluster/tests/test_virtual_cluster.py

```diff
@@ -0,0 +1,405 @@
+    webui_url = format_web_url(webui_url)
+
+    # Add two nodes to the primary cluster.
+    cluster.add_node(env_vars={"RAY_NODE_TYPE_NAME": "4c8g"})
+    cluster.add_node(env_vars={"RAY_NODE_TYPE_NAME": "8c16g"})
+    hostname = socket.gethostname()
+
+    virtual_cluster_id = ""
+    allocation_mode = ""
+    replica_sets = {}
+    revision = 0
+
+    def _create_or_update_virtual_cluster():
+        try:
+            nonlocal revision
+            resp = requests.post(
+                webui_url + "/virtual_clusters",
+                json={
+                    "virtualClusterId": virtual_cluster_id,
+                    "allocationMode": allocation_mode,
+                    "replicaSets": replica_sets,
+                    "revision": revision,
+                },
+            )
+            resp.raise_for_status()
+            result = resp.json()
+            print(result)
+            assert result["result"] is True, resp.text
+            assert result["data"]["virtualClusterId"] == virtual_cluster_id
+            current_revision = result["data"]["revision"]
+            assert current_revision > revision
+            revision = current_revision
+            virtual_cluster_replica_sets = {}
+            for _, node_instance in result["data"]["nodeInstances"].items():
+                assert node_instance["hostname"] == hostname
+                virtual_cluster_replica_sets[node_instance["templateId"]] = (
+                    virtual_cluster_replica_sets.get(node_instance["templateId"], 0) + 1
+                )
+            # The virtual cluster has the same node types and count as expected.
+            assert replica_sets == virtual_cluster_replica_sets
+            return True
+        except Exception as ex:
+            logger.info(ex)
+            return False
+
+    # Create a new virtual cluster with exclusive allocation mode.
+    virtual_cluster_id = "virtual_cluster_1"
+    allocation_mode = "exclusive"
+    replica_sets = {"4c8g": 1, "8c16g": 1}
+    wait_for_condition(_create_or_update_virtual_cluster, timeout=10)
+
+    # Update the virtual cluster with less nodes.
+    replica_sets = {"4c8g": 1}
+    wait_for_condition(_create_or_update_virtual_cluster, timeout=10)
+
+    # Update the virtual cluster with more nodes.
+    replica_sets = {"4c8g": 1, "8c16g": 1}
+    wait_for_condition(_create_or_update_virtual_cluster, timeout=10)
+
+    # Update the virtual cluster with zero node (make it empty).
+    replica_sets = {}
+    wait_for_condition(_create_or_update_virtual_cluster, timeout=10)
+
+    # `virtual_cluster_1` has released all nodes, so we can now
+    # create a new virtual cluster with two nodes.
+    virtual_cluster_id = "virtual_cluster_2"
+    replica_sets = {"4c8g": 1, "8c16g": 1}
+    # Using mixed allocation mode.
+    allocation_mode = "mixed"
+    wait_for_condition(_create_or_update_virtual_cluster, timeout=10)
+
+    # TODO: uncomment this when `MixedCluster::IsIdleNodeInstance` is fixed.
```

Comment:
We have to fix `MixedCluster::IsIdleNodeInstance`, making it work with `ClusterResourceManager`